interface/ceed-basis.c

```diff
@@ -1303,82 +1307,89 @@ int CeedSymmetricSchurDecomposition(Ceed ceed, CeedScalar *mat,
   for (CeedInt i=0; i<n-1; i++) {
     // Calculate Householder vector, magnitude
     CeedScalar sigma = 0.0;
-    v[i] = matT[i+n*(i+1)];
+    v[i] = mat_T[i+n*(i+1)];
     for (CeedInt j=i+1; j<n-1; j++) {
-      v[j] = matT[i+n*(j+1)];
+      v[j] = mat_T[i+n*(j+1)];
       sigma += v[j] * v[j];
     }
     CeedScalar norm = sqrt(v[i]*v[i] + sigma); // norm of v[i:n-1]
-    CeedScalar Rii = -copysign(norm, v[i]);
-    v[i] -= Rii;
+    CeedScalar R_ii = -copysign(norm, v[i]);
+    v[i] -= R_ii;
     // norm of v[i:m] after modification above and scaling below
     //   norm = sqrt(v[i]*v[i] + sigma) / v[i];
     //   tau = 2 / (norm*norm)
-    if (sigma > 10*CEED_EPSILON)
+    if (sigma > CEED_EPSILON)
       tau[i] = 2 * v[i]*v[i] / (v[i]*v[i] + sigma);
     else
       tau[i] = 0;
 
+    // Don't apply Householder reflector if row already diagonal
+    if (sigma < CEED_EPSILON) continue;
+
+    mat_T[i+n*(i+1)] = R_ii;
+    mat_T[(i+1)+n*i] = R_ii;
+
     for (CeedInt j=i+1; j<n-1; j++)
       v[j] /= v[i];
 
     // Update sub and super diagonal
-    matT[i+n*(i+1)] = Rii;
-    matT[(i+1)+n*i] = Rii;
     for (CeedInt j=i+2; j<n; j++) {
-      matT[i+n*j] = 0; matT[j+n*i] = 0;
+      mat_T[i+n*j] = 0; mat_T[j+n*i] = 0;
     }
     // Apply symmetric Householder reflector to lower right panel
-    CeedHouseholderReflect(&matT[(i+1)+n*(i+1)], &v[i], tau[i],
+    CeedHouseholderReflect(&mat_T[(i+1)+n*(i+1)], &v[i], tau[i],
                            n-(i+1), n-(i+1), n, 1);
-    CeedHouseholderReflect(&matT[(i+1)+n*(i+1)], &v[i], tau[i],
+    CeedHouseholderReflect(&mat_T[(i+1)+n*(i+1)], &v[i], tau[i],
                            n-(i+1), n-(i+1), 1, n);
+
     // Save v
     for (CeedInt j=i+1; j<n-1; j++) {
-      matT[i+n*(j+1)] = v[j];
+      mat_T[i+n*(j+1)] = v[j];
     }
   }
   // Backwards accumulation of Q
   for (CeedInt i=n-2; i>=0; i--) {
-    v[i] = 1;
-    for (CeedInt j=i+1; j<n-1; j++) {
-      v[j] = matT[i+n*(j+1)];
-      matT[i+n*(j+1)] = 0;
+    if (tau[i] > CEED_EPSILON) {
+      v[i] = 1;
+      for (CeedInt j=i+1; j<n-1; j++) {
+        v[j] = mat_T[i+n*(j+1)];
+        mat_T[i+n*(j+1)] = 0;
+      }
+      CeedHouseholderReflect(&mat[(i+1)+n*(i+1)], &v[i], tau[i],
+                             n-(i+1), n-(i+1), n, 1);
```

Comment:
What is this conditional intended to do? If `v[]` contains large entries, then `tau` would be very small, and yet skipping the reflection doesn't seem right. Note that `tau v^T v = 2` so if v has size 1e10, then tau has size 2e-20.

Reply:
This is the one that was added as an effort to reduce noise. I pushed a change.